File: TypeScript/PaymentsService/src/Infrastructure/Payments/PayPal/PayPalProcessPaymentRequestValidator.ts
import { IProcessPaymentRequestValidator } from "../../../Domain/Payments/IProcessPaymentRequestValidator";
import { ProcessPaymentBaseRequest } from "../../../Domain/Payments/ProcessPaymentBaseRequest";
import { RequestValidationResult } from "../../../Domain/Payments/RequestValidationResult";
import { ProcessPayPalPaymentRequest } from "./ProcessPayPalPaymentRequest";

export class PayPalProcessPaymentRequestValidator implements IProcessPaymentRequestValidator {
    private constructor() { }

    private static instance: PayPalProcessPaymentRequestValidator;

    public static get Instance(): PayPalProcessPaymentRequestValidator {
        if(!this.instance) {
            this.instance = new PayPalProcessPaymentRequestValidator()
        }

        return this.instance;
    }

    public validateRequest(request: ProcessPaymentBaseRequest): RequestValidationResult {
        const req = request as ProcessPayPalPaymentRequest;

        if(!req) {
            return new RequestValidationResult(false, "BadRequest");
        }

        if (!req.token) {
            return new RequestValidationResult(false, 'BadToken');
        }

        return new RequestValidationResult(true);
    }
}

File: CSharp/PaymentsService/Infrastructure/Payments/PayPal/PayPalProcessPaymentRequestValidator.cs
using PaymentsService.Domain.Payments;

namespace PaymentsService.Infrastructure.Payments.PayPal;

public sealed class PayPalProcessPaymentRequestValidator : IProcessPaymentRequestValidator
{
    private PayPalProcessPaymentRequestValidator() { }

    private static readonly Lazy<PayPalProcessPaymentRequestValidator> _lazyInstance = new(() => new PayPalProcessPaymentRequestValidator());
    public static PayPalProcessPaymentRequestValidator Instance => _lazyInstance.Value;

    public RequestValidationResult ValidateRequest(ProcessPaymentBaseRequest request)
    {
        if (request is not ProcessPayPalPaymentRequest req)
        {
            return new RequestValidationResult(false, "BadRequest");
        }

        if (string.IsNullOrEmpty(req.Token))
        {
            return new RequestValidationResult(false, "BadToken");
        }

        return new RequestValidationResult(true);
    }
}
